### PR TITLE
Apply data source v2 changes

### PR DESCRIPTION
**Pull request by mccheah**
Applies the series of patches to get the latest data source v2 API, particularly for the table catalog and v2 logical plans.

### CODE REVIEW

**Comment by mccheah** [comment]
@robert3005 @vinooganesh @yifeih

**Comment by vinooganesh** [comment]
@mccheah should we just take Scala 2.11 rather than do the hacks?

File: project/MimaExcludes.scala [hunk 1]
```diff
@@ -165,7 +165,84 @@ object MimaExcludes {
...
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.Filter"),
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.IsNull"),
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.EqualTo"),
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.And$"),
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.Or$"),
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.StringStartsWith"),
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.StringContains$"),
+
+    // [SPARK-26216][SQL] Do not use case class as public API (UserDefinedFunction)
+    ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.expressions.UserDefinedFunction$"),
```
Comment (robert3005) [inline project/MimaExcludes.scala]: You likely need to revert changes below

**Comment by robert3005** [comment]
I think if you revert the sql parser changes then this is fine

**Pull request closed without merging**